### ISSUE TITLE
Memory seems to not be release between PDF generations in production environment

### ISSUE THREAD

**Issue by leandrobattochio** [QuestPDF/QuestPDF#958]
I'm generating a PDF from a RabbitMQ queue. In my consumer handler, I get all the info I need and in the end I call GeneratePdf() to get the bytes[], then I upload the contents to S3 and the logic is finished.

When I'm testing localhost, memory consumption goes high until the pdf is generated, and then stays that way. It does not decrease, it seems nothing is freed from the memory. If I publish another message in the queue, the consumer starts generating the Pdf again but in a given moment, memory is released the processing begins again.

So the graphic for two generations in a row is something like this:

![image](https://github.com/user-attachments/assets/789f57a4-775e-4611-bfca-8eeb11ad9439)

My pdf has around 1,2k pages with more than 180 images. Final memory consumption is somemthing like 2gb.

![image](https://github.com/user-attachments/assets/52881534-6eff-40a6-8968-34f973839e75)

Images are downloaded outside QuestPDF components.

When I run this code in production environment in AWS, I get a different result. Memory is not released like in this graphic, so every time I generate a new PDF, it ends up with just a little bit more memory than what it used before.

I sense something is wrong here. What can I do to investigate this further? Since I get the expected behaviour in my local machine, and I'm running the same code in prod, I'm a little bit confused here.

**Comment by MarcinZiabek** [comment]
Hi 😄 Thank you for sharing this observation.

Do you know if this behavior has remained consistent since you opened this issue? Is there anything else to share? The server GC may behave differently that the localhost GC.

Could you please briefly describe the content of your document? Number of pages, number of images, number of text instances, etc.

If the problem persists, could you please downgrade to the 204.6.X version?

Thank you for your understanding and collaboration.

**Comment by leandrobattochio** [comment]
Document content:

- Around 1200 pages;
- Around 200 images that gets resized;
- It contains A LOT of tables, charts and text. I cannot say exactly how much text instances I have, but from 1200 pages I'd say 1100 is full with texts and tables and when not charts or images.

Using version 2024.6.0 I get this graph:
![image](https://github.com/user-attachments/assets/49f740d3-bfa5-46e0-8d10-6d9f533f589e)
Just after the generation, seems like some memory is released but not all. Then if I call again, memory drop and start rising again just before I starting the PDF generation.

![image](https://github.com/user-attachments/assets/ac355a01-a519-4541-91f7-18f6b6d3773e)

After the second generation, not memory is released like in the first time. I will try to repeat the same tests in production when I have a chance.

Memory snapshot taken after 2nd generation localhost: (memory suddenly dropped after the snapshot)
![image](https://github.com/user-attachments/assets/caa694d6-13fe-4e90-8ec1-d5d20110355c)

![image](https://github.com/user-attachments/assets/c6098b49-cde2-4346-9da1-56c7a06968cb)

Does it help?

**Comment by danirzv** [comment]
@leandrobattochio did I guess you get unmanaged memory on creating `MemoryStream` did you try `RecyclableMemoryStream`?
https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream

**Comment by kerams** [comment]
The byte arrays behind `MemoryStream` are `RecyclableMemoryStream` are all managed. However, you're right that there seems to be quite a lot of unreleased unmanaged memory, perhaps suggesting a leak in the QuestPDF <-> native Skia layer.

**Comment by danirzv** [comment]
> The byte arrays behind `MemoryStream` are `RecyclableMemoryStream` are all managed. However, you're right that there seems to be quite a lot of unreleased unmanaged memory, perhaps suggesting a leak in the QuestPDF <-> native Skia layer. Or maybe the handles for unmanaged resources are just kept around for future reuse.

If he is using `RecyclableMemoryStream` and dispose it right, probably unmanaged memory should get released, but i need to investigate more if you are suspected to 'QuestPDF <-> native Skia layer'

**Comment by MarcinZiabek** [comment]
@kerams Could you please prepare a minimal example of what causes the suspected memory leak? I have tested many scenarios related to native memory, and I think all of the issues have been fixed. Of course, I might have missed something. 

From my personal observations, on MacOS, the native memory usage graph in Rider is total garbage and completely does not correspond to the real memory usage of an application shown in the Activity Monitor. I hope that your tools are more accurate and we can rely on them during testing/experimenting.

Could you please share an example of your document? So I will understand its size, complexity, etc.? You can send it here or via contact@questpdf.com

**Comment by leandrobattochio** [comment]
> @leandrobattochio did I guess you get unmanaged memory on creating `MemoryStream` did you try `RecyclableMemoryStream`? https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream

I think I didn't understand what you meant with "did you try with RecyclableMemoryStream?".

This is how I'm generating the PDF:

`
    public async Task<IDocument> Generate(DadosRelatorioHelper relatorio, DadosEmpresaHelper dadosEmpresa)
    {
        logger.LogDebug("Starting Pdf Generation {Kind} {ReportId}", relatorio.ReportType, relatorio.Id);

        IDocument? capaDocument = null;

        var images = await ProcessImages(dadosEmpresa);

        var createCapaTask = Task.Run(() => { capaDocument = GenerateCapa(relatorio, images); });

        IDocument? relatorioDocument = null;
        var taskCreatRelatorio = Task.Run(() =>
        {
            logger.LogDebug("Generating Report content");
            switch (relatorio.ReportType)
            {
                case CompanyReportTypeEnum.AET:
                {
                    var data = new RelatorioAETData(relatorio, dadosEmpresa);
                    relatorioDocument = relatorioAET.GenerateDocument(data, images);
                    break;
                }
               (... ommited switch ... )
            }
        });

        await Task.WhenAll([taskCreatRelatorio, createCapaTask]);
        
        if (relatorioDocument != null && capaDocument != null)
        {
            logger.LogDebug("Merging cover with content");

            var m = Document.Merge(capaDocument, relatorioDocument)
                .UseContinuousPageNumbers();
            return m;
        }

        logger.LogError("I should not be here");
        return null!;
    }
`

This is inside a Mediator handler, that means, after I get the bytes from GeneratePdf, I upload it to S3, update my database with the S3 link and return. something like:


`
    public async Task<CreatePdfCommandResult> Handle(CreatePdfCommand request,
        CancellationToken cancellationToken = default)
    {
        _logger.LogInformation("CreatePdfHandler called - Starting PDF generation {ReportId}",
            request.GenerationRequest.ReportId);

        ( ... ommited code ... getting information from database )
        try
        {
                var document = await Generate(dadosRelatorio.Adapt<DadosRelatorioHelper>(),
                    dadosEmpresa.Adapt<DadosEmpresaHelper>());

                var rawData = document.GeneratePdf();
                
                if (rawData == null)
                {
                    _logger.LogError("rawData when generating PDF was null. {ReportId}",
                        request.GenerationRequest.ReportId);

                    relatorio.SetFailure(); // PdfStatus = 3
                    return new CreatePdfCommandResult(PdfStatusEnum.Falha, "failure generating");
                }

               (... ommited - upload and save link ... )

                relatorio.SetPdfGeneratedSuccess(); // PdfStatus = 2
                return new CreatePdfCommandResult(PdfStatusEnum.PdfGerado, relatorio.Descricao);
            }
            catch (Exception ex)
            {
                relatorio.SetFailure(); // PdfStatus = 3

                _logger.LogError(ex, "Failure generatind PDF {ReportId}", request.GenerationRequest.ReportId);
                return new CreatePdfCommandResult(PdfStatusEnum.Falha, "failure generating");
            }
            finally
            {
                 (... save and persist changes Entity Framework ...)
            }
        }

        return new CreatePdfCommandResult(PdfStatusEnum.Falha, "I shouldn't be here");
    }
`

**Comment by sps-campbellwray** [comment]
I suspect that these two issues might be related: https://github.com/QuestPDF/QuestPDF/issues/968

 @leandrobattochio does the 2024.7.2 release make a difference for you in your development environment? i.e. does the memory go back down immediately once the generation is finished, as opposed to only dropping when the second one starts?

**Comment by MarcinZiabek** [comment]
Generally speaking, additional memory will be consumed after generating the first document: the library loads and caches fonts and other structures in the native memory. Additionally, we are comparing desktop and server garbage collectors, which may use different algorithms for memory management. 

I am always open to investigating and addressing memory leaks and other performance problems. I am just saying it is expected to see specific differences between localhost and production environments.

If you experience any memory leak, any help in locating its place and providing more details like document characteristics, would be highly helpful 😄

**Comment by leandrobattochio** [comment]
What more details specifically do you need other than what I just said about the document? I'll be glad to provide them.

**Comment by MarcinZiabek** [comment]
Any form of a minimal repository that allows me to reproduce this issue would be highly appreciated. I am delighted to fix any problems related to the library; I just need something to analyze. 😄

**Comment by sps-campbellwray** [comment]
@MarcinZiabek I'm not sure if @leandrobattochio's issue is specifically related to images (he did mention that his PDF has 200 images, and I believe that all QuestPDF images use `SkBitmap` under the hood), but I created a fork of QuestPDF a few weeks back that is able to reproduce the memory leak when generating images from the PDF, which might be helpful as a starting point.

He also mentioned that:
> The sudden memory drop is when my second PDF generation started in localhost. Seems like good behaviour.

> When I run this code in production environment in AWS, I get a different result. Memory is not released like in this graphic, so every time I generate a new PDF, it ends up with just a little bit more memory than what it used before.

Which is somewhat consistent with what I was seeing, where everything works fine on Local, but in Docker the memory is not being released. It wouldn't surprise me if this was a bug related to Linux, which is why we are both seeing it, but only in Production.

You can find the repo I created here [QuestPDF.MemoryLeak](https://github.com/sps-campbellwray/QuestPDF.MemoryLeak/tree/main/Source), and you can find more information on how to test it in my [comment](https://github.com/QuestPDF/QuestPDF/issues/968#issuecomment-2276456837)

**Comment by MarcinZiabek** [comment]
Thank you so much for sharing this minimal example. I have carefully reviewed both managed and native code related to generating images. So far, I think that the QuestPDF implementation seems to be correct. 

I suspect a memory leak on the Skia side. The `2024.7.3` has been recently released, which bumps the Skia dependency from m126 to m128 and possibly upgrades many other dependencies along the way.

As it is pretty easy to do, could you please try the latest QuestPDF release? Also, if possible, please use the latest dotnet 8 version to increase our confidence in memory usage metrics.

**Comment by MarcinZiabek** [comment]
Code related to this issue:
- https://github.com/QuestPDF/QuestPDF/blob/main/Source/QuestPDF/Drawing/ImageCanvas.cs
- https://github.com/QuestPDF/QuestPDF.Native/blob/main/managed/NativeSkia/SkBitmap.cs
- https://github.com/QuestPDF/QuestPDF.Native/blob/main/native/src/bitmap.cpp
- https://github.com/QuestPDF/QuestPDF.Native/blob/main/managed/NativeSkia/SkData.cs
- https://github.com/QuestPDF/QuestPDF.Native/blob/main/native/src/data.cpp

**Comment by JanHergenhan** [comment]
Hi there! 😊

We also have encountered the issue with memory not being released in our container environment.

Our service is running in a Docker container (mcr.microsoft.com/dotnet/runtime:6.0) on Kubernetes in AWS. The container has 512 MB of memory allocated. We are generating PDFs using the document.GeneratePdf() method, without generating any images.

I conducted some tests with different versions of a 5-page PDF, both with and without images. In both cases, we observed that memory is not being sufficiently released over time, leading to the container running out of memory when a sufficient number of requests are made. The chart below illustrates the memory usage of our two containers. In the first half, we sent 6 requests per minute and generated PDFs with images. In the second half, we increased the request rate to 11 per minute, but the PDFs did not contain any images. As a result, we experience memory exhaustion every 8 hours.
![image](https://github.com/user-attachments/assets/f9d32667-cd2d-4641-90cf-507e6d76add4)

While there are ways to mitigate this issue, our preference would be to identify and resolve the memory leak.

I have looked into the code you suggested. However, I have limited experience with unmanaged resources in .NET. During my research, I came across two potential improvements (maybe you already considered/discarded them):

- Using SafeHandle instead of IntPtr because it offers better cleanup capabilities. see [here](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle?view=net-8.0).
- Implementing the dispose pattern instead of a plain Dispose() implementation. see [here](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose).

I am tempted to give these improvements a try when I find the time (although it won't be possible in the next two weeks). Therefore, I wanted to inquire if there is any documentation available on how to build QuestPDF.Native, including the native sources for SKIA, and how to incorporate them into the QuestPDF NuGet package. @MarcinZiabek I would appreciate any guidance you can provide on this matter. Thanks!

**Comment by MarcinZiabek** [comment]
@JanHergenhan 

Could you please share your minimal example? 

What environment are you using? What is the library version you have used for testing?

Does the memory usage come back to normal after some time? (e.g. GC begins work once higher memory pressure is achieved) Or do you need to restart the microservice to make it operational?

**Comment by MarcinZiabek** [comment]
@JanHergenhan 

At this moment, I think the issue could be simply related to the garbage collector and the difference between workstation and server modes. When I force server GC by using `<ServerGarbageCollection>true</ServerGarbageCollection>` in csproj, the GC minimizes the frequency of collecting operations for the cost of higher (up to 650 MB) memory usage.

Could your server be memory limited (e.g., 512 MB), and the default server GC mode does not run optimally under such constraint?

**Comment by JanHergenhan** [comment]
@MarcinZiabek 
I cannot check exactly at the moment but it was one of the latest release candidates, probably 2024.10.0-rc4. The environment is this container: mcr.microsoft.com/dotnet/runtime:6.0, based on Debian 11, the exact runtime version should be 6.0.34. 

The container is limited to 512 MB of memory so we could run into the scenario you mentioned. We could give it some more memory, we're not that limited.

The memory usage only goes down in tiny steps from time to time but never enough. The drop of memory usage in the graph I showed comes from the container being forcefully restarted due to exceeding the allowed memory. The x-axis is actually to coarse to actually see that we reach 100%.

**Comment by JanHergenhan** [comment]
@MarcinZiabek 
Ok, I can confirm that we were running QuestPDF.2024.10.0-rc4 in .NET6

As you suggested, I ran the tests also with `<ServerGarbageCollection>false</ServerGarbageCollection>` to enforce the workstation mode. However, the results are quite similar:
![image](https://github.com/user-attachments/assets/ca1ecb75-86fa-4411-88f0-a8d81adcac2d)

I'm sending 11 requests per minute to generate the pdf with images, memory exhaustion occurrs every ~7 hours. So it seems that GC works slightly better in workstation mode (previous tests had 6 requests per minute and exhaustion after ~8 hours) but we're still running out of memory over time...

**Comment by kerams** [comment]
Are you on ARM by any chance?

**Comment by JanHergenhan** [comment]
We actually are.

**Comment by kerams** [comment]
I might also be seeing this on ARM in production, although I am generating few PDFs at this time, so no OOM issues yet. It's been a while, but I think when I tried to create a repro benchmark on x86, I did not get anywhere.

**Comment by JanHergenhan** [comment]
@MarcinZiabek 
so I ran the tests also locally on my developer machine (Win11), where the application is running in the Visual Studio debugger with Postman sending periodic requests.

I am by far no expert in analyzing memory in VS and do not know how to export the data so here is a screenshot of the result:
![image](https://github.com/user-attachments/assets/456a1c66-6551-458b-905d-fa29c4eef1fa)

The memory consumption is constantly increasing while the gabage collector does only very little. The heap analysis shows a constant increase of `SkParagraphBuilder` or rather the `ConcurrentBag` objects. The never seem to be cleaned up properly. Btw, a high number of `SkParagraph` objects is also visible in the last image of the original post.

However, looking at the numbers, their impact on (managed) memory seems small. Not sure what they might be hiding in the native memory...

**Comment by MarcinZiabek** [comment]
Thank you all for being active in this discussion. I highly appreciate your input!

@JanHergenhan Based on your screenshot, I recognize that the `SkParagraphBuilder` objects are unnecessarily created. The functionality has been implemented as an object pool, and instances should be shared and reused across `Text` element instances and documents. 

There is a separate pool per paragraph style, which consists of text alignment, text direction, maximum lines visible, and line-clamp text. Is there any chance that you are dynamically setting those values, e.g., during any measurement in the `dynamic` component? For example, when looking for a perfect document structure. I would be looking for any loops, etc. If you can share your QuestPDF code, that would help me a lot.

I am primarily testing MacOS M2. I will continue the investigation focused on the abnormally high number of `SkParagraphBuilder`, which could be the source of the issue.

**Comment by JanHergenhan** [comment]
@MarcinZiabek 

I sent you the code via mail to contact@questpdf.com.

I'm using `Placeholder`s in my code to simulate varying data (mostly addresses). There are also many single `Text` elements in the document. I'd not expect that this creates lots of paragraph styles as there is not a lot of variance in style. Not sure what might cause the many different keys in the `ConcurrentDictionary`. I might be doing something in the wrong way (static Typography class, usage of `Component`s....), it's my first shot with QuestPdf.

I hope you can reproduce it on your machine. Thank you :)

**Comment by MarcinZiabek** [comment]
I think I have found a root cause. I am still unable to reproduce the issue, but my finding aligns nicely with the fact the problem appears on specific CPU architectures and operating systems 😄 

Could you please try the 2024.10.2 version?

**Comment by JanHergenhan** [comment]
Thanks a lot, I'll try it out and come back with results :)

After looking at your changes, I'm wondering what you actually mean with "my finding aligns nicely with the fact the problem appears on specific CPU architectures and operating systems".

**Comment by MarcinZiabek** [comment]
All `SkParagraphBuilder` instances were stored and shared between documents and text elements inside multiple object pools. For each distinct paragraph configuration, one pool was created. Pools were stored in a dictionary, where the key was a record with all paragraph properties. However, one of those properties is the text of the ellipsis. But instead of using the string itself, it used a pointer value (IntPtr) to the memory location where the native text value was stored. It turns out the native memory is quite stable on Windows and MacOS, and the address happens to be the same all the time. But in certain cases, I imagine that the native text (byte array) could be created in various places of the native memory, depending on the memory allocation typical to the processor architecture and operating system. And different address results with a different key value for the dictionary, and a separate pool with a single item. This could explain why you have observed so many pools (concurrent bags) and builder instances, of similar count.

I am quite positive I have nailed it 😄  (hopefully this time I got lucky)

**Comment by JanHergenhan** [comment]
@MarcinZiabek 

first of all, this is the memory of our service from the last 12 hours:
![image](https://github.com/user-attachments/assets/bde8a954-6b68-40ce-8c4c-6e847c58a73a)
There is a marginal increase over time but we're now far away from "out of memory every 8h" 🎉 Thanks a lot!


Regarding native memory: The screenshot from earlier https://github.com/QuestPDF/QuestPDF/issues/958#issuecomment-2431363002 is from a Windows machine with an Intel processor (not the ARM docker container). So in this case, I'd then not expect those many instances of pools?!

Btw: I ran the code I sent you also in a console application in a loop and there the garbage collection worked actually fine. So I assume that the memory leak is also related to ASP.NET Core and maybe the dependency injection (the generation runs in a scoped service which is called from a controller). So there is obviously something blocking a proper cleanup, not only of QuestPDF objects.

I have only very little experience with GC in .NET so take those assumptions with a big grain of salt. My knowledge on GC stems from the stuff I read on that topic in the last days [(parts of this)](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#gc-fundamentals)

I had a look into the QuestPDF.Native package. Your objects implement the IDisposable interface. However, you're usually calling the Dispose method from within the Finalizers. I'm not conviced that this is the way to do that because you heavily rely on the finalizers to actually get called.

![image](https://github.com/user-attachments/assets/9e990e46-e633-4ed4-a2ad-f82ec92b82f5)
In this screenshot (dev machine, but v2024.10.2), I checked the "show dead objects" checkbox with the yellow triangle. As far as I understand, those are objects ready for finalization. And they're piling up, possibly because the finalizer thread doesn't run. When I force finalization with `GC.WaitForPendingFinalizers()' from time to time (as @sps-campbellwray did in his fork), those dead objects are actually removed and the memory consumption drops.

**tl;dr:** It works now, big thank you again. I'm just not sure if the root cause might still be there. I'm lacking experience (and time) to investigate further and will happily continue using your library :)

**Comment by MarcinZiabek** [comment]
Not quite 😄 

The vast majority of objects are disposed of before the finalizer is called. Finalizers are used as a last-resort option to mitigate risks related to memory management. 

The only place I can think of that relies on finalizers is handling images, which are likely the biggest in size objects allocated natively. An alternative solution would be to traverse the entire document after the generation process, look for unmanaged resources and free them up. In my measurements, memory pressure does not change much. 

However, our latest experiences show that native allocation and garbage collection can occasionally surprise us. It might be worth investigating.

Thank you for sharing crucial tips and observations that have allowed us to resolve this problem.

Important: if anyone observes similar memory-related problems, please create a new issue.

**Comment by MarcinZiabek** [comment]
In Q1 2025, a significant effort was made to address memory leak issues and optimise memory consumption. Since Q2 2025, we have not received any reports related to memory leak problems, so we are closing this issue. :confirm 

I would like to thank everybody involved in the investigation and testing - your help in making QuestPDF stable and reliable has been invaluable! ❤️
